web/src/settings_data.ts

```diff
@@ -370,35 +371,40 @@ export function get_request_data_for_stream_privacy(selected_val: string): {
     history_public_to_subscribers: boolean;
     is_web_public: boolean;
 } {
+    // When changing a channel from public to private with public history,
+    // "history_public_to_subscribers" value is same and does not change.
+    // So, it will not be included in the data returned by
+    // "populate_data_for_stream_settings_request" as it only includes
+    // the field whose new value has changed from its original value.
+    //
+    // So, we need to include "history_public_to_subscribers" here
+    // itself so that the correct value is passed to the API, otherwise
+    // private streams are set to have private history by default.
+    const history_public_to_subscribers: boolean = $("#id_history_public_to_subscribers").is(
+        ":checked",
+    );
     switch (selected_val) {
         case settings_config.stream_privacy_policy_values.public.code: {
             return {
                 is_private: false,
-                history_public_to_subscribers: true,
+                history_public_to_subscribers,
                 is_web_public: false,
             };
         }
         case settings_config.stream_privacy_policy_values.private.code: {
             return {
                 is_private: true,
-                history_public_to_subscribers: false,
+                history_public_to_subscribers,
                 is_web_public: false,
             };
         }
-        case settings_config.stream_privacy_policy_values.web_public.code: {
+        default: {
```

Comment:
This is probably better written with `case settings_config.stream_privacy_policy_values.web_public.code` and throwing an exception in the default case, if we think that shouldn't be possible, right?

Reply:
Maybe there's something fancy we could do with ensuring the type is such that TypeScript knows there aren't other possibilities, but it doesn't seem important.